piff/gsobject_model.py

```diff
@@ -0,0 +1,389 @@
+from .util import hsm
+
+
+class GSObjectModel(Model):
+    """ Model that takes a fiducial GalSim.GSObject and dilates, shifts, and shears it to get a
+    good match to stars.
+
+    :param gsobj:    GSObject to use as fiducial profile.
+    :param fastfit:  Use HSM moments for fitting.  Approximate, but fast.  [default: False]
+    :param force_model_center: If True, PSF model centroid is fixed at origin and
+                        PSF fitting will marginalize over stellar position.  If False, stellar
+                        position is fixed at input value and the fitted PSF may be off-center.
+                        [default: True]
+    :param include_pixel:   Include integration over pixel?  [default: True]
+    :param logger:   A logger object for logging debug info. [default: None]
+    """
+    def __init__(self, gsobj, fastfit=False, force_model_center=True, include_pixel=True,
+                 logger=None):
+        import galsim
+        if isinstance(gsobj, basestring):
+            gsobj = eval(gsobj)
+
+        self.kwargs = {'gsobj':repr(gsobj),
+                       'fastfit':fastfit,
+                       'force_model_center':force_model_center,
+                       'include_pixel':include_pixel}
+
+        self.gsobj = gsobj
+        self._fastfit = fastfit
+        self._force_model_center = force_model_center
+        self._method = 'auto' if include_pixel else 'no_pixel'
+        # Params are [u, v], size, g1, g2
+        if self._force_model_center:
+            self._nparams = 3
+            params = [1.0, 0.0, 0.0]
+        else:
+            self._nparams = 5
+            params = [0.0, 0.0, 1.0, 0.0, 0.0]
+
+        # Calibrate gsobj by measuring it with HSM.  This way we can use differences in HSM moments
+        # to get a reasonable starting guess for stars.
+        prof = self.getProfile(params)
+        img = prof.drawImage(method=self._method)
+        sd = StarData(img, img.trueCenter())
+        fiducial_star = Star(sd, None)
+        flux, cenu, cenv, size, g1, g2, flag = hsm(fiducial_star)
+        if flag != 0:
+            raise RuntimeError("Error calculating fiducial moments for this gsobject.")
+        self._fiducial_size = size
+        self._fiducial_shape = galsim.Shear(g1=g1, g2=g2)
+        self._fiducial_flux = flux
+        self._fiducial_centroid = galsim.PositionD(cenu, cenv)
+
+    def moment_fit(self, star, logger=None):
+        """Estimate transformations needed to bring self.gsobj towards given star."""
+        import galsim
+        flux, cenu, cenv, size, g1, g2, flag = hsm(star)
+        centroid = galsim.PositionD(cenu, cenv)
+        shape = galsim.Shear(g1=g1, g2=g2)
+
+        if flag:
+            raise RuntimeError("Failed measuring HSM moments of star.")
+
+        dflux = flux/self._fiducial_flux
+        dcentroid = centroid - self._fiducial_centroid
+        dsize = size/self._fiducial_size
+        dshape = shape - self._fiducial_shape
```

Comment:
I think all this fiducial stuff isn't working the way you want.  I was investigating why the tests in test_single_image (in test_simple.py) lost 1 d.p.  You use fastfit there, which should be equivalent to the old code, so the agreement shouldn't have changed.

I think I tracked it down to this bit of code.  Here you return the values relative to the fiducial values, but AFAICT you don't ever re-apply the fiducual values to get the real size, g1, g2, etc.  Furthermore, it looks like just a slightly lower precision because the fiducial flux, size, g1,g2 are in this case: 1.0000512599945068, 1.0000552542315129, -7.9303995699720791e-17, -9.589749686788564e-18.  So the deltas are not much different from the true values.

So I think there are probably two bugs here.  First, it seems like there is something wrong with how you are computing the fiducial values if this is what comes back for them.  And second, there are presumably places where you need to be adding or multiplying them back in to get the values you actually care about to send to the interpolator.

Or possibly, you may just want to scrap the fiducial stuff rather than fix it, since it's apparently not doing anything right now other than mess up the 5th decimal place.  Which means the code would probably work fine (or better) without it.